### PR TITLE
Replace some lodash methods by native methods [WIP]

### PR DESCRIPTION
**Pull request by Haroenv**
This should be tested thoroughly!

35.1 KB --> 34.7 KB

### CODE REVIEW

**Comment by Haroenv** [comment]
Are there only 5 tests?

**Comment by Haroenv** [comment]
arg, I forgot that _.map and _.reduce also work on objects; should check if that's the case here...

File: src/SearchParameters/RefinementList.js [hunk 1]
```diff
@@ -104,7 +103,7 @@ var lib = {
     } else if (isFunction(attribute)) {
       var hasChanged = false;
 
-      var newRefinementList = reduce(refinementList, function(memo, values, key) {
+      var newRefinementList = refinementList.reduce(function(memo, values, key) {
```
Comment (Haroenv) [inline src/SearchParameters/RefinementList.js]: This might be an Object 🤔

File: src/SearchParameters/index.js [hunk 2]
```diff
@@ -894,7 +890,7 @@ SearchParameters.prototype = {
       return omit(this.numericRefinements, attribute);
     } else if (isFunction(attribute)) {
       var hasChanged = false;
-      var newNumericRefinements = reduce(this.numericRefinements, function(memo, operators, key) {
+      var newNumericRefinements = this.numericRefinements.reduce(function(memo, operators, key) {
```
Comment (Haroenv) [inline src/SearchParameters/index.js]: object maybe?

File: src/SearchParameters/index.js [hunk 3]
```diff
@@ -518,9 +514,9 @@ SearchParameters._parseNumbers = function(partialState) {
     forEach(partialState.numericRefinements, function(operators, attribute) {
       numericRefinements[attribute] = {};
       forEach(operators, function(values, operator) {
-        var parsedValues = map(values, function(v) {
-          if (isArray(v)) {
-            return map(v, function(vPrime) {
+        var parsedValues = values.map(function(v) {
```
Comment (Haroenv) [inline src/SearchParameters/index.js]: object maybe

File: src/SearchParameters/index.js [hunk 4]
```diff
@@ -1499,7 +1495,7 @@ SearchParameters.prototype = {
     return intersection(
       // enforce the order between the two arrays,
       // so that refinement name index === hierarchical facet index
-      map(this.hierarchicalFacets, 'name'),
+      this.hierarchicalFacets.map('name'),
```
Comment (Haroenv) [inline src/SearchParameters/index.js]: object

File: src/SearchParameters/index.js [hunk 5]
```diff
@@ -1693,7 +1689,7 @@ SearchParameters.prototype = {
       this.getHierarchicalFacetByName(facetName)
     );
     var path = refinement.split(separator);
-    return map(path, trim);
+    return path.map(trim);
```
Comment (Haroenv) [inline src/SearchParameters/index.js]: probably an object

File: src/SearchResults/generate-hierarchical-tree.js [hunk 6]
```diff
@@ -31,7 +29,7 @@ function generateTrees(state) {
       results = hierarchicalFacetResult.slice(hierarchicalRootPath.split(hierarchicalSeparator).length);
     }
 
-    return reduce(results, generateTreeFn, {
+    return results.reduce(generateTreeFn, {
```
Comment (Haroenv) [inline src/SearchResults/generate-hierarchical-tree.js]: maybe an object

File: src/SearchResults/index.js [hunk 7]
```diff
@@ -334,7 +332,7 @@ function SearchResults(state, results) {
    * disjunctive facets results
    * @member {SearchResults.HierarchicalFacet[]}
    */
-  this.hierarchicalFacets = map(state.hierarchicalFacets, function initFutureTree() {
+  this.hierarchicalFacets = state.hierarchicalFacets.map(function initFutureTree() {
```
Comment (Haroenv) [inline src/SearchResults/index.js]: object?

File: src/SearchResults/index.js [hunk 8]
```diff
@@ -510,7 +508,7 @@ function SearchResults(state, results) {
     });
   });
 
-  this.hierarchicalFacets = map(this.hierarchicalFacets, generateHierarchicalTree(state));
+  this.hierarchicalFacets = this.hierarchicalFacets.map(generateHierarchicalTree(state));
```
Comment (Haroenv) [inline src/SearchResults/index.js]: object?

File: src/SearchResults/index.js [hunk 9]
```diff
@@ -545,7 +543,7 @@ function extractNormalizedFacetValues(results, attribute) {
     var facet = find(results.facets, predicate);
     if (!facet) return [];
 
-    return map(facet.data, function(v, k) {
+    return facet.data.map(function(v, k) {
```
Comment (Haroenv) [inline src/SearchResults/index.js]: object?

File: src/SearchResults/index.js [hunk 10]
```diff
@@ -557,7 +555,7 @@ function extractNormalizedFacetValues(results, attribute) {
     var disjunctiveFacet = find(results.disjunctiveFacets, predicate);
     if (!disjunctiveFacet) return [];
 
-    return map(disjunctiveFacet.data, function(v, k) {
+    return disjunctiveFacet.data.map(function(v, k) {
```
Comment (Haroenv) [inline src/SearchResults/index.js]: object?

File: src/SearchResults/index.js [hunk 11]
```diff
@@ -578,7 +576,7 @@ function recSort(sortFn, node) {
   if (!node.data || node.data.length === 0) {
     return node;
   }
-  var children = map(node.data, partial(recSort, sortFn));
+  var children = node.data.map(partial(recSort, sortFn));
```
Comment (Haroenv) [inline src/SearchResults/index.js]: object?

File: src/algoliasearch.helper.js [hunk 12]
```diff
@@ -1184,7 +1183,7 @@ AlgoliaSearchHelper.prototype._search = function() {
 
   this.emit('search', state, this.lastResults);
 
-  var derivedQueries = map(this.derivedHelpers, function(derivedHelper) {
+  var derivedQueries = this.derivedHelpers.map(function(derivedHelper) {
```
Comment (Haroenv) [inline src/algoliasearch.helper.js]: probably an array but maybe object?

File: src/requestBuilder.js [hunk 13]
```diff
@@ -251,8 +248,7 @@ var requestBuilder = {
   _getHitsHierarchicalFacetsAttributes: function(state) {
     var out = [];
 
-    return reduce(
-      state.hierarchicalFacets,
+    return state.hierarchicalFacets.reduce(
```
Comment (Haroenv) [inline src/requestBuilder.js]: probably object

File: src/SearchResults/index.js (hunk 8 above)
Comment (vvo) [reply src/SearchResults/index.js]: > arg, I forgot that _.map and _.reduce also work on objects; should check if that's the case here...

Object.entries() works well for those cases

Comment (Haroenv) [reply src/SearchResults/index.js]: Yep, will have to use object.keys in those (entries is edge, end helper supports ie9)

Comment (vvo) [reply src/SearchResults/index.js]: This should be discussed with everyone about the browser support of helper, IE9 is really old. We can rely on polyfills. Like InstantSearch does

Comment (Haroenv) [reply src/SearchResults/index.js]: I’d rather not add polyfills in a PR aimed at making the size smaller, but we can go further in removing methods if ie11 or edge is picked as target

File: src/SearchParameters/RefinementList.js (hunk 1 above)
Comment (iam4x) [reply src/SearchParameters/RefinementList.js]: Object here, look the reducer function 👍

File: src/SearchParameters/index.js (hunk 2 above)
Comment (iam4x) [reply src/SearchParameters/index.js]: Yes object as well

File: src/SearchParameters/index.js (hunk 4 above)
Comment (iam4x) [reply src/SearchParameters/index.js]: This should be:

```js
this.hierarchicalFacets.map(({name}) => name)
```

File: src/SearchParameters/index.js (hunk 5 above)
Comment (iam4x) [reply src/SearchParameters/index.js]: Nope, `refinement.split` would return an array 👍

File: src/SearchResults/generate-hierarchical-tree.js (hunk 6 above)
Comment (iam4x) [reply src/SearchResults/generate-hierarchical-tree.js]: Nope it's an array, worked on that piece 👍

File: src/SearchResults/index.js (hunk 7 above)
Comment (iam4x) [reply src/SearchResults/index.js]: Array as well

**Comment by bobylito** [comment]
Hi @Haroenv and thanks a lot for your contribution and reopening the discussion. The move to native methods on collections has been a question for a long time now. This was never because of a few key questions that we had:
 - what is the size benefit that we can expect?
 - what is the impact on the dev experience?
 - what is the impact on performance?
 - can we actually replace lodash completely and at what cost?

Can you go in further details in your analysis on those topics? Maybe you can figure out all the lodash functions used (and based on their usage) and see the one that have zero cost for removal (eg: map for arrays), small (eg: map for objects / collections)  and big (function that don't have a native counterpart).

Also on the perf part, Lodash has always been a library that strives for perf. What is the cost of going native? Are there any browsers where this impact is worse?

The objectives here must be that:
 - the size gain must be substantial
 - we don't want to support a lodash-like
 - the dev experience should not be impacted (it's already complex enough)

**Comment by bobylito** [comment]
Hi @Haroenv and thanks a lot for your contribution and reopening the discussion. The move to native methods on collections has been a question for a long time now. This was never because of a few key questions that we had:
 - what is the size benefit that we can expect?
 - what is the impact on the dev experience?
 - what is the impact on performance?
 - can we actually replace lodash completely and at what cost?

Can you go in further details in your analysis on those topics? Maybe you can figure out all the lodash functions used (and based on their usage) and see the one that have zero cost for removal (eg: map for arrays), small (eg: map for objects / collections)  and big (function that don't have a native counterpart).

Also on the perf part, Lodash has always been a library that strives for perf. What is the cost of going native? Are there any browsers where this impact is worse?

The objectives here must be that:
 - the size gain must be substantial
 - we don't want to support a lodash-like
 - the dev experience should not be impacted (it's already complex enough)

File: src/SearchResults/index.js (hunk 8 above)
Comment (Haroenv) [reply src/SearchResults/index.js]: For the record: `this.hierarchicalFacets` is an array

**Comment by Haroenv** [comment]
Let's discuss this in #552 instead. We can easily replicate the work done here later

**Pull request closed without merging**